### PR TITLE
Triangle numerical integration

### PR DESCRIPTION
**Pull request by hdeling**
This pull-request adds the following features 
- add numerical integration on tetrahedra and triangles
- add a mechanical tag on a PointSetGeometry object in order to find its associated MechanicalObject
- add the orthogonal of a 2d vector
- add support of high order elements in VTK and Gmesh mesh loaders

It also fixes some bugs 
-Quand2Triangle Topologivcal mapping

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: SofaKernel/framework/sofa/core/loader/MeshLoader.cpp [hunk 1]
```diff
@@ -39,10 +39,15 @@ MeshLoader::MeshLoader() : BaseLoader()
     , d_triangles(initData(&d_triangles,"triangles","Triangles of the mesh loaded"))
     , d_quads(initData(&d_quads,"quads","Quads of the mesh loaded"))
     , d_polygons(initData(&d_polygons,"polygons","Polygons of the mesh loaded"))
+    , highOrderEdgePositions(initData(&highOrderEdgePositions,"highOrderEdgePositions","High order edge points of the mesh loaded"))
```
Comment (ChristianDuriez) [inline SofaKernel/framework/sofa/core/loader/MeshLoader.cpp]: Hi Hervé !
Now we have naming conventions in SOFA:
https://github.com/sofa-framework/sofa/blob/master/GUIDELINES.md
see Rule N5 in Naming section: d_ for data

File: SofaKernel/framework/sofa/core/loader/MeshLoader.h [hunk 2]
```diff
@@ -93,12 +111,17 @@ class SOFA_CORE_API MeshLoader : public BaseLoader
     Data< helper::vector< Triangle > > d_triangles;
     Data< helper::vector< Quad > > d_quads;
     Data< helper::vector< helper::vector <unsigned int> > > d_polygons;
+    Data< helper::vector< HighOrderEdgePosition > > highOrderEdgePositions;
```
Comment (ChristianDuriez) [inline SofaKernel/framework/sofa/core/loader/MeshLoader.h]: Here also, naming data convention d_

File: SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h [hunk 3]
```diff
@@ -132,6 +128,8 @@ class PointSetGeometryAlgorithms : public core::topology::GeometryAlgorithms
     sofa::core::topology::BaseMeshTopology* m_topology;
     Data<float> showIndicesScale;
     Data<bool> showPointIndices;
+    /// Tage of the Mechanical State associated with the vertex position
+    Data<std::string> m_tagMechanics;
```
Comment (ChristianDuriez) [inline SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h]: d_tagMechanics;

m_ is for other variable names [N7]

File: modules/SofaTopologyMapping/Quad2TriangleTopologicalMapping.cpp [hunk 4]
```diff
@@ -99,10 +99,10 @@ void Quad2TriangleTopologicalMapping::init()
             toModel->setNbPoints(fromModel->getNbPoints());
 
 
-            for (int i=0; i<fromModel->getNbPoints(); i++)
-            {
-                toModel->addPoint(fromModel->getPX(i), fromModel->getPY(i), fromModel->getPZ(i));
-            }
+            //for (int i=0; i<fromModel->getNbPoints(); i++)
```
Comment (damienmarchal) [inline modules/SofaTopologyMapping/Quad2TriangleTopologicalMapping.cpp]: Shouldn't the commented code removed ?

File: SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h [hunk 5]
```diff
@@ -132,6 +128,8 @@ class PointSetGeometryAlgorithms : public core::topology::GeometryAlgorithms
     sofa::core::topology::BaseMeshTopology* m_topology;
     Data<float> showIndicesScale;
```
Comment (ChristianDuriez) [inline SofaKernel/modules/SofaBaseTopology/PointSetGeometryAlgorithms.h]: These data could also be renamed

File: SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TriangleNumericalIntegration_test.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    }
+	bool testNumericalIntegration()
+	{
+		size_t k;
+		Real integral,weight;
+		typename NumericalIntegrationDescriptor::BarycentricCoordinatesType bc;
+		Vec<3,unsigned short> randomPolynomial;
+
+		// get the descriptor of numerical integration on tetrahedra
+        NumericalIntegrationDescriptor &nid=geo->getTriangleNumericalIntegrationDescriptor();
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TriangleNumericalIntegration_test.cpp]: Identation looks weird.

**Comment by damienmarchal** [comment]
Hi @hdeling,

Thank you for your PR. 
At first glance it looks nice only cosmetic/minor details.

File: modules/SofaGeneralEngine/GenerateGrid.inl [hunk 7]
```diff
@@ -54,16 +52,14 @@ GenerateGrid<DataTypes>::GenerateGrid()
...
 void GenerateGrid<DataTypes>::init()
 {
-    addInput(&f_length);
-    addInput(&f_height);
-    addInput(&f_width);
-    addInput(&f_origin);
-    addInput(&f_resolutionLength);
-    addInput(&f_resolutionWidth);
-    addInput(&f_resolutionHeight);
+    addInput(&f_minCorner);
```
Comment (ChristianDuriez) [inline modules/SofaGeneralEngine/GenerateGrid.inl]: If you can change all the data from
f_ 
to 
d_ 
it would be great ;-))

**Review by ChristianDuriez — CHANGES_REQUESTED** [review]
Thanks Hervé ! 
Very useful changes.  I didn't really get if it can be used for computation of internal force in Triangle or Tetrahedron ForceFields ?

Only the same change requested
put d_  before datatypes 
(see N5 at https://github.com/sofa-framework/sofa/blob/master/GUIDELINES.md)

File: modules/SofaTopologyMapping/Quad2TriangleTopologicalMapping.cpp (hunk 4 above)
Comment (hdeling) [reply modules/SofaTopologyMapping/Quad2TriangleTopologicalMapping.cpp]: Yes of course.

**Review by ChristianDuriez — APPROVED** [review]

**Comment by guparan** [comment]
Should we revert .gitignore modification in cfa5a44 ?

**Comment by hdeling** [comment]
Yes probably should remove the commit on .gitignore.

**Review by BenjaminGilles — APPROVED** [review]
thanks Hervé, this looks all good to me.
minor comment:  some functions might be moved from  SofaBaseTopology/CommonAlgorithms.h  to defaulttype/Vec.h

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
Hi all,

It seems there a lot of new test-scene failures since the merge of this PR. 
https://ci.inria.fr/sofa-ci/job/centos_clang-3.4_options/635/console

We probably need to fix that in a new PR.

**Pull request merged**